controller/handler_edge_ctrl/heartbeat.go

```diff
@@ -38,17 +39,35 @@ func (h *sessionHeartbeatHandler) ContentType() int32 {
 	return env.ApiSessionHeartbeatType
 }
 
-func (h *sessionHeartbeatHandler) HandleReceive(msg *channel2.Message, _ channel2.Channel) {
+func (h *sessionHeartbeatHandler) HandleReceive(msg *channel2.Message, ch channel2.Channel) {
 	go func() {
 		req := &edge_ctrl_pb.ApiSessionHeartbeat{}
+		routerId := ch.Id().Token
 		if err := proto.Unmarshal(msg.Body, req); err == nil {
 
-			err := h.appEnv.GetHandlers().ApiSession.MarkActivity(req.Tokens)
+			notFoundTokens, err := h.appEnv.GetHandlers().ApiSession.MarkActivity(req.Tokens)
 
 			if err != nil {
 				pfxlog.Logger().
 					WithError(err).
-					Error("unable to set activity for heartbeat")
+					WithField("routerId", routerId).
+					Error("unable to set activity for heartbeat: %v", err)
+			}
+
+			if len(notFoundTokens) > 0 {
+				pfxlog.Logger().
+					WithField("routerId", routerId).
+					Debugf("api session tokens not found during heartbeat, sending delete: %v", notFoundTokens)
+
+				msgStruct := &edge_ctrl_pb.ApiSessionRemoved{
+					Tokens: notFoundTokens,
+				}
+
+				msgBytes, _ := proto.Marshal(msgStruct)
+				msg := channel2.NewMessage(env.ApiSessionRemovedType, msgBytes)
+
+				//if we don't delete it this time, we will try again when we see it
```

Comment:
I would just use Send. The timeout only adds overhead if you're not checking for an error

Reply:
I was trying to avoid blocking as this is an unbounded go routine generator.

Reply:
Send will return faster than SendWithTimeout. Send will return as soon as it's queued. SendWithTimeout will wait until it's off the queue and written to network, timing out if it doesn't make it onto the network in time. 

SendWithTimeout won't start the timeout until the message is queued for send.